src/main/java/de/tum/in/www1/artemis/service/TextAssessmentQueueService.java

```diff
@@ -0,0 +1,158 @@
+            }
+            if (textBlock.getCluster() == null) {
+                continue;
+            }
+            double textBlockScore = textBlock.getAddedDistance();
+            textBlockScore /= textBlock.getCluster().size();
+            textBlockScore += smallerClusterMap.get(textBlock);
+            totalScore += textBlockScore;
+        }
+        return totalScore;
+    }
+
+    /**
+     * Sums up all the distances in the Blocks cluster
+     *
+     * @param textBlock textBlock for which the distance should be added
+     * @param cluster in which the textBlock distance should be added up
+     * @throws IllegalArgumentException if textBlock isn't an element of cluster
+     * @return added Distance to all other textblocks in the cluster
+     */
+    private double calculateAddedDistance(TextBlock textBlock, TextCluster cluster) {
+        if (!cluster.getBlocks().contains(textBlock)) {
+            throw new IllegalArgumentException("textBlock must be an element of the cluster");
+        }
+        double[][] distanceMatrix = cluster.getDistanceMatrix();
+        int blockID = cluster.getBlocks().indexOf(textBlock);
+        // subtract 1 because the statement also included the distance to itself, but it should't be included
+        return Arrays.stream(distanceMatrix[blockID]).map(distance -> 1.0 - distance).sum() - 1;
+    }
+
+    /**
+     * Calculates and sets all AddedDistances to a list of TextBlocks in a Cluster
+     *
+     * @param textBlockList list of the TextBlocks
+     * @param textCluster Cluster
+     */
+    public void setAddedDistances(List<TextBlock> textBlockList, TextCluster textCluster) {
+        textBlockList.forEach(textBlock -> {
+            double addedDistance = calculateAddedDistance(textBlock, textCluster);
+            textBlock.setAddedDistance(addedDistance);
+        });
+    }
+
+    /**
+     * Calculates the Percentages of Smaller Clusters for a list of textSubmissions
+     *
+     * @param textSubmissionList for which the smaller clusters should be calculated
+     * @throws IllegalArgumentException if not all TextSubmissions are from the same exercise
+     * @return return a HashMap where the textBlock is the key and smaller cluster percentage is the value
+     * If a textBlock has no cluster or is already assessable, it isn't in the HashMap
+     */
+    public HashMap<TextBlock, Double> calculateSmallerClusterPercentageBatch(List<TextSubmission> textSubmissionList) {
+        HashMap<TextBlock, Double> result = new HashMap<>();
+        if (textSubmissionList.isEmpty()) {
+            return result;
+        }
+        TextExercise currentExercise = (TextExercise) textSubmissionList.get(0).getParticipation().getExercise();
+        List<TextCluster> clusters = textClusterRepository.findAllByExercise(currentExercise);
+
+        if (!textSubmissionList.stream().map(Submission::getParticipation).map(Participation::getExercise).allMatch(elem -> elem == currentExercise)) {
```

Comment:
Comparing the Exercise Id might save unnecessary Joins and Hibernate Lazy Loading the same Exercise Element.

Reply:
No Databasecalls are necessary here